File: backend/app/models/phieu_tra_hang.py
from datetime import date, datetime, timezone
from decimal import Decimal
from sqlalchemy import Date, DateTime, ForeignKey, Integer, Numeric, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.database import Base


class PhieuTraHang(Base):
    """Phiếu khách trả hàng — phôi (PHOI) hoặc thành phẩm (THANH_PHAM)."""
    __tablename__ = "phieu_tra_hang"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    so_phieu: Mapped[str] = mapped_column(String(30), unique=True, nullable=False)
    ngay: Mapped[date] = mapped_column(Date, nullable=False)
    loai_hang: Mapped[str] = mapped_column(String(20), nullable=False)
    # PHOI | THANH_PHAM
    customer_id: Mapped[int] = mapped_column(
        Integer, ForeignKey("customers.id"), nullable=False
    )
    production_order_id: Mapped[int | None] = mapped_column(
        Integer, ForeignKey("production_orders.id"), nullable=True
    )
    delivery_order_id: Mapped[int | None] = mapped_column(
        Integer, ForeignKey("delivery_orders.id"), nullable=True
    )
    warehouse_id: Mapped[int] = mapped_column(
        Integer, ForeignKey("warehouses.id"), nullable=False
    )
    ly_do_tra: Mapped[str | None] = mapped_column(Text)
    trang_thai: Mapped[str] = mapped_column(String(20), nullable=False, server_default="draft")
    # draft | confirmed | huy
    nguoi_giao: Mapped[str | None] = mapped_column(String(100))
    ghi_chu: Mapped[str | None] = mapped_column(Text)
    created_by: Mapped[int | None] = mapped_column(Integer, ForeignKey("users.id"))
    confirmed_by: Mapped[int | None] = mapped_column(Integer, ForeignKey("users.id"))
    confirmed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
    )

    customer = relationship("Customer")
    production_order = relationship("ProductionOrder")
    delivery_order = relationship("DeliveryOrder")
    warehouse = relationship("Warehouse")
    creator = relationship("User", foreign_keys=[created_by])
    confirmer = relationship("User", foreign_keys=[confirmed_by])
    items: Mapped[list["PhieuTraHangItem"]] = relationship(
        "PhieuTraHangItem", cascade="all, delete-orphan", back_populates="phieu"
    )


class PhieuTraHangItem(Base):
    __tablename__ = "phieu_tra_hang_items"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    phieu_id: Mapped[int] = mapped_column(
        Integer, ForeignKey("phieu_tra_hang.id", ondelete="CASCADE"), nullable=False
    )
    so_luong: Mapped[int] = mapped_column(Integer, nullable=False)
    don_vi: Mapped[str | None] = mapped_column(String(20))  # Tấm | Thùng | ...
    tinh_trang: Mapped[str] = mapped_column(String(10), nullable=False, server_default="tot")
    # tot | loi
    chieu_kho: Mapped[Decimal | None] = mapped_column(Numeric(8, 2))   # PHOI only
    chieu_cat: Mapped[Decimal | None] = mapped_column(Numeric(8, 2))   # PHOI only
    product_id: Mapped[int | None] = mapped_column(
        Integer, ForeignKey("products.id"), nullable=True
    )  # THANH_PHAM only
    don_gia: Mapped[Decimal | None] = mapped_column(Numeric(18, 2))
    ghi_chu: Mapped[str | None] = mapped_column(Text)

    phieu: Mapped["PhieuTraHang"] = relationship(
        "PhieuTraHang", back_populates="items"
    )
    product = relationship("Product")

File: frontend/src/api/phieu_tra_hang.ts
import api from './client'

export type LoaiHang = 'PHOI' | 'THANH_PHAM'

export interface TraHangItem {
  id?: number
  so_luong: number
  don_vi?: string | null
  tinh_trang: 'tot' | 'loi'
  chieu_kho?: number | null    // PHOI only
  chieu_cat?: number | null    // PHOI only
  product_id?: number | null   // THANH_PHAM only
  ten_san_pham?: string | null
  don_gia?: number | null
  thanh_tien?: number | null
  ghi_chu?: string | null
}

export interface PhieuTraHang {
  id: number
  so_phieu: string
  ngay: string
  loai_hang: LoaiHang
  customer_id: number
  ten_khach_hang: string | null
  production_order_id: number | null
  so_lenh: string | null
  delivery_order_id: number | null
  warehouse_id: number
  ten_kho: string | null
  ly_do_tra: string | null
  trang_thai: 'draft' | 'confirmed' | 'huy'
  nguoi_giao: string | null
  ghi_chu: string | null
  created_at: string | null
  confirmed_at: string | null
  tong_so_luong: number
  tong_tot: number
  tong_loi: number
  items: TraHangItem[]
}

export interface PhieuTraHangCreate {
  ngay: string
  loai_hang: LoaiHang
  customer_id: number
  production_order_id?: number | null
  delivery_order_id?: number | null
  warehouse_id: number
  ly_do_tra?: string | null
  nguoi_giao?: string | null
  ghi_chu?: string | null
  items: TraHangItem[]
}

export interface PhieuTraHangListParams {
  loai_hang?: LoaiHang
  customer_id?: number
  production_order_id?: number
  trang_thai?: string
  tu_ngay?: string
  den_ngay?: string
}

const BASE = '/phieu-tra-hang'

export const phieuTraHangApi = {
  list: (params?: PhieuTraHangListParams) =>
    api.get<PhieuTraHang[]>(BASE, { params }).then(r => r.data),

  get: (id: number) =>
    api.get<PhieuTraHang>(`${BASE}/${id}`).then(r => r.data),

  create: (body: PhieuTraHangCreate) =>
    api.post<PhieuTraHang>(BASE, body).then(r => r.data),

  update: (id: number, body: Partial<PhieuTraHangCreate>) =>
    api.put<PhieuTraHang>(`${BASE}/${id}`, body).then(r => r.data),

  delete: (id: number) =>
    api.delete(`${BASE}/${id}`),

  confirm: (id: number) =>
    api.post<PhieuTraHang>(`${BASE}/${id}/confirm`).then(r => r.data),

  huy: (id: number) =>
    api.post(`${BASE}/${id}/huy`).then(r => r.data),
}
